block creation of StorageClass until prerequisites are met

umangachapagain opened the pull request:
CephBlockPool waits for CephCluster to be ready.
Similarly, StorageClass needs to wait for CephBlockPool to be ready.

Same for CephFilesystem.

jarrpa reviewed (changes requested):
Unless I'm missing something, this doesn't seem to take into account that either of CephBlockPool or CephFilesystem could be disabled. Would it make more sense to extend `StorageClusterConfiguration` to include a "required kind" or something so that we can just do the check on a per-SC basis here? https://github.com/openshift/ocs-operator/blob/master/controllers/storagecluster/storageclasses.go#L88-L93

umangachapagain commented:
> Unless I'm missing something, this doesn't seem to take into account that either of CephBlockPool or CephFilesystem could be disabled. Would it make more sense to extend `StorageClusterConfiguration` to include a "required kind" or something so that we can just do the check on a per-SC basis here? https://github.com/openshift/ocs-operator/blob/master/controllers/storagecluster/storageclasses.go#L88-L93
>
Don't we always create both the StorageClasses?

I don't think a "required kind" is necessary since we know what is required.
>
> Also, CI failures look valid. Check `make ocs-operator-ci` output.
>
Yes, they are valid and I give up on fixing it. It requires re-factoring entire unit test framework (which is actually running integration test internally).
Unless someone has a better way to fix it.

jarrpa commented:
> > Unless I'm missing something, this doesn't seem to take into account that either of CephBlockPool or CephFilesystem could be disabled. Would it make more sense to extend `StorageClusterConfiguration` to include a "required kind" or something so that we can just do the check on a per-SC basis here? https://github.com/openshift/ocs-operator/blob/master/controllers/storagecluster/storageclasses.go#L88-L93
> 
> Don't we always create both the StorageClasses?

No. If there's no CephBlockPool created it makes no sense to create a StorageClass pointing to something that doesn't exist. 

> I don't think a "required kind" is necessary since we know what is required.

It would be a cleaner implementation. Rather than having an `if/else` reading the StorageClass name and checking for the appropriate Kind based on that, We could just read the same field for all configs. But... eh, you're right, it's not strictly necessary, so I won't block on it.

> > Also, CI failures look valid. Check `make ocs-operator-ci` output.
> 
> Yes, they are valid and I give up on fixing it. It requires re-factoring entire unit test framework (which is actually running integration test internally).
> Unless someone has a better way to fix it.

The `ocs-operator-ci` fixes are straightforward linting problems, so please at least deal with that. :P

The e2e failures look like CI infra issues, so we just need to retest. That can wait until LGTM.

umangachapagain commented:
> > > Unless I'm missing something, this doesn't seem to take into account that either of CephBlockPool or CephFilesystem could be disabled. Would it make more sense to extend `StorageClusterConfiguration` to include a "required kind" or something so that we can just do the check on a per-SC basis here? https://github.com/openshift/ocs-operator/blob/master/controllers/storagecluster/storageclasses.go#L88-L93
> 
Updated to check if particular StorageClass is enabled before trying to wait. But, it seems like StorageClass creation doesn't care about the setting anyway and that's a different PR.

> > > Also, CI failures look valid. Check `make ocs-operator-ci` output.
>
Fixed the annoying uppercase/lowercase issue.

umangachapagain commented:
/retest

jarrpa reviewed (approved):
ONE. MORE. TIME.

/lgtm

jarrpa commented:
/override ci/prow/ci-index-dev-master-dependencies
